---
layout: post
title: Introduction to Promise
subtitle: Standard built-in objects
date: 2019-04-17
author: Jalever
header-img: img/post-bg-js-version.jpg
catalog: true
tags:
  - JavaScript
---
- [Overview](#overview)
- [Syntax](#syntax)
    - [Parameters](#parameters)
- [Description](#description)
- [Properties](#properties)
    - [Promise.length](#promiselength)
    - [Promise.prototype](#promiseprototype)
- [Methods](#methods)
    - [Promise.all(iterable)](#promisealliterable)
    - [Promise.allSettled()](#promiseallSettled)
    - [Promise.race(iterable)](#promiseraceiterable)
    - [Promise.reject()](#promisereject)
    - [Promise.resolve()](#promiseresolve)
- [Promise prototype](#promise-prototype)
    - [Properties](#properties)
        - [Promise.prototype.constructor](#promiseprototypeconstructor)
    - [Methods](#methods)
        - [Promise.prototype.catch()](#promiseprototypecatch)
        - [Promise.prototype.then()](#promiseprototypethen)
        - [Promise.prototype.finally()](#promiseprototypefinally)
- [Create a Promise](#create-a-promise)    
- [Examples](#examples)
    - [Basic Example](#basic-example)

## Overview
The Promise object represents the eventual completion (or failure) of an asynchronous operation, and its resulting value.
![e9Clin.png](https://s2.ax1x.com/2019/07/21/e9Clin.png)

## Syntax
![e9CBJ1.png](https://s2.ax1x.com/2019/07/21/e9CBJ1.png)

#### Parameters
&nbsp;&nbsp;<strong>executor</strong><br/>
&nbsp;&nbsp;A function that is passed with the arguments `resolve` and `reject`. The `executor` function is executed immediately by the Promise implementation, passing `resolve` and `reject` functions (the executor is called before the `Promise` constructor even returns the created object). The `resolve` and `reject` functions, when called, resolve or reject the promise, respectively. The executor normally initiates some asynchronous work, and then, once that completes, either calls the `resolve` function to resolve the promise or else `rejects` it if an error occurred. If an error is thrown in the executor function, the promise is rejected. The return value of the executor is ignored.

## Description
A `Promise` is a proxy for a value not necessarily known when the promise is created. It allows you to associate handlers with an asynchronous action's eventual success value or failure reason. This lets asynchronous methods return values like synchronous methods: instead of immediately returning the final value, the asynchronous method returns a <i>Promise</i> to supply the value at some point in the future.

A `Promise` is in one of these states:
- <strong>pending</strong>: initial state, neither fulfilled nor rejected.
- <strong>fulfilled</strong>: meaning that the operation completed successfully.
- <strong>rejected</strong>: meaning that the operation failed.

A pending promise can either be <strong>fulfilled</strong> with a value, or <strong>rejected</strong> with a reason (error). When either of these options happens, the associated handlers queued up by a promise's `then` method are called. (If the promise has already been fulfilled or rejected when a corresponding handler is attached, the handler will be called, so there is no race condition between an asynchronous operation completing and its handlers being attached.)

As the <strong>Promise.prototype.then()</strong> and <strong>Promise.prototype.catch()</strong> methods return promises, they can be chained.
![e9AWG9.png](https://s2.ax1x.com/2019/07/21/e9AWG9.png)
> A Promise is said to be settled if it is either `fulfilled` or `rejected`, but not pending. You will also hear the term `resolved` used with promises — this means that the promise is settled or “locked in” to match the state of another promise.

## Properties
#### Promise.length
Length property whose value is always 1 (number of constructor arguments).

#### Promise.prototype  
The `Promise.prototype` property represents the prototype for the `Promise` constructor.

`Promise` instances inherit from `Promise.prototype`. You can use the constructor's prototype object to add properties or methods to all `Promise` instances.

<table>
    <thead>
        <tr>
            <td colspan="3">Property attributes of Promise.prototype</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>Writable</td>
            <td>no</td>
        </tr>
        <tr>
            <td>Enumerable</td>
            <td>no</td>
        </tr>
        <tr>
            <td>Configurable</td>
            <td>no</td>
        </tr>
    </tbody>
</table>

## Methods
#### Promise.all(iterable)
The `Promise.all()` method returns a single `Promise` that resolves when all of the promises passed as an iterable have resolved or when the iterable contains no promises. It rejects with the reason of the first promise that rejects.
![e9JkYn.png](https://s2.ax1x.com/2019/07/21/e9JkYn.png)

###### Syntax
![e9J0kd.png](https://s2.ax1x.com/2019/07/21/e9J0kd.png)

&nbsp;&nbsp;<strong>Parameters</strong><br/>
&nbsp;&nbsp;&nbsp;&nbsp;<strong>iterable</strong><br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An iterable object such as an Array.<br/>
&nbsp;&nbsp;<strong>Return value</strong><br/>
&nbsp;&nbsp;&nbsp;&nbsp;- An <strong>already resolved</strong> `Promise` if the iterable passed is empty.
&nbsp;&nbsp;&nbsp;&nbsp;- An <strong>asynchronously resolved</strong> `Promise` if the iterable passed contains no promises. Note, Google Chrome 58 returns an <strong>already resolved</strong> promise in this case.
&nbsp;&nbsp;&nbsp;&nbsp;- A <strong>pending</strong> `Promise` in all other cases. This returned promise is then resolved/rejected <strong>asynchronously</strong> (as soon as the stack is empty) when all the promises in the given iterable have resolved, or if any of the promises reject. See the example about "Asynchronicity or synchronicity of Promise.all" below. Returned values will be in order of the Promises passed, regardless of completion order.

###### Description
This method can be useful for aggregating the results of multiple promises.

<strong>Fulfillment</strong><br/>
The returned promise is fulfilled with an array containing all the values of the <i>iterable</i> passed as argument (also non-promise values).
- If an empty <i>iterable</i> is passed, then this method returns (synchronously) an already resolved promise.
- If all of the passed-in promises fulfill, or are not promises, the promise returned by <strong>Promise.all</strong> is fulfilled asynchronously.

<strong>Rejection</strong><br/>
If any of the passed-in promises reject, <strong>Promise.all</strong> asynchronously rejects with the value of the promise that rejected, whether or not the other promises have resolved.

###### Examples
<strong>Using Promise.all</strong><br/>
Promise.all waits for all fulfillments (or the first rejection).
![e9UGUe.png](https://s2.ax1x.com/2019/07/21/e9UGUe.png)
If the <ins>iterable</ins> contains non-promise values, they will be ignored, but still counted in the returned promise array value (if the promise is fulfilled):
![e9Ug8s.png](https://s2.ax1x.com/2019/07/21/e9Ug8s.png)
![e9UfK0.png](https://s2.ax1x.com/2019/07/21/e9UfK0.png)

<strong>Asynchronicity or synchronicity of Promise.all</strong><br/>
This following example demonstrates the asynchronicity (or synchronicity, if the iterable passed is empty) of <strong>Promise.all</strong>:
![e9dkY4.png](https://s2.ax1x.com/2019/07/21/e9dkY4.png)

The same thing happens if <strong>Promise.all</strong> rejects:
![e9wEE8.png](https://s2.ax1x.com/2019/07/21/e9wEE8.png)

But, <strong>Promise.all</strong> resolves synchronously <ins>if and only if</ins> the <i>iterable</i> passed is empty:
![e90nIO.png](https://s2.ax1x.com/2019/07/21/e90nIO.png)

<strong>Promise.all fail-fast behaviour</strong><br/>
<strong>Promise.all</strong> is rejected if any of the elements are rejected. For example, if you pass in four promises that resolve after a timeout and one promise that rejects immediately, then <strong>Promise.all</strong> will reject immediately.
![e90vSH.png](https://s2.ax1x.com/2019/07/21/e90vSH.png)
![e9BSOI.png](https://s2.ax1x.com/2019/07/21/e9BSOI.png)
It is possible to change this behaviour by handling possible rejections:
![e90IyR.png](https://s2.ax1x.com/2019/07/21/e90IyR.png)

#### Promise.allSettled()

#### Promise.race(iterable)

#### Promise.reject()

#### Promise.resolve()

## Promise prototype
#### Properties
###### Promise.prototype.constructor

#### Methods
###### Promise.prototype.catch()

###### Promise.prototype.then()

###### Promise.prototype.finally()

## Create a Promise
A Promise object is created using the new keyword and its constructor. This constructor takes as its argument a function, called the "executor function". This function should take two functions as parameters. The first of these functions (resolve) is called when the asynchronous task completes successfully and returns the results of the task as a value. The second (reject) is called when the task fails, and returns the reason for failure, which is typically an error object.
![e9nIsA.png](https://s2.ax1x.com/2019/07/21/e9nIsA.png)
To provide a function with promise functionality, simply have it return a promise:
![e9n7Zt.png](https://s2.ax1x.com/2019/07/21/e9n7Zt.png)

## Examples
#### Basic Example
![e9MQ2D.png](https://s2.ax1x.com/2019/07/21/e9MQ2D.png)
